src/middlewared/middlewared/plugins/account_/privilege.py

```diff
@@ -0,0 +1,161 @@
+        return item
+
+    @accepts(Patch(
+        "privilege_entry",
+        "privilege_create",
+        ("rm", {"name": "builtin_name"}),
+    ))
+    async def do_create(self, data):
+        """
+        Creates a privilege.
+
+        `local_groups` is a list of local user account groups that gain this privilege.
+
+        `ds_groups` is list of Directory Service group IDs that will gain this privilege.
+        """
+        await self._validate("privilege_create", data)
+
+        id = await self.middleware.call(
+            "datastore.insert",
+            self._config.datastore,
+            data
+        )
+
+        return self.get_instance(id)
+
+    @accepts(
+        Int("id", required=True),
+        Patch(
+            "privilege_entry",
+            "privilege_update",
+            ("rm", {"name": "builtin_name"}),
+            ("attr", {"update": True}),
+        )
+    )
+    async def do_update(self, id, data):
+        """
+        Update the privilege `id`.
+        """
+        old = await self.get_instance(id)
+        new = old.copy()
+        new["local_groups"] = [g["id"] for g in new["local_groups"]]
+        new.update(data)
+
+        verrors = ValidationErrors()
+
+        if new["builtin_name"]:
+            for k in ["name", "allowlist"]:
+                if new[k] != old[k]:
+                    verrors.add(f"privilege_update.{k}", "This field is read-only for built-in privileges")
+
+            builtin_privilege = BuiltinPrivileges(new["builtin_name"])
+
+            if builtin_privilege == BuiltinPrivileges.LOCAL_ADMINISTRATOR:
+                if not await self.middleware.call("group.has_password_enabled_user", new["local_groups"]):
+                    verrors.add(
+                        "privilege_update.local_groups",
+                        "None of the members of these groups has password login enabled. At least one grantee of "
+                        "the \"Local Administrator\" privilege must have password login enabled."
+                    )
+
+        verrors.check()
+
+        new.update(data)
+
+        await self._validate("privilege_update", new, id)
+
+        await self.middleware.call(
+            "datastore.update",
+            self._config.datastore,
+            id,
+            new,
+        )
+
+        return self.get_instance(id)
+
+    @accepts(
+        Int("id")
+    )
+    async def do_delete(self, id):
+        """
+        Delete the privilege `id`.
+        """
+        privilege = await self.get_instance(id)
+        if privilege["builtin_name"]:
+            raise CallError("Unable to delete built-in privilege", errno.EPERM)
+
+        response = await self.middleware.call(
+            "datastore.delete",
+            self._config.datastore,
+            id
+        )
+
+        return response
+
+    async def _validate(self, schema_name, data, id=None):
+        verrors = ValidationErrors()
```

Comment:
Are we doing any validation on the user-provided `resources`?

Reply:
As for API key, currently no. No values there could harm us in any way.

Reply:
But how do we tell user they've fat-fingered something?

Reply:
Nohow yet